src/gravel/controllers/nodes/deployment.py

```diff
@@ -265,3 +288,79 @@ async def join(
 
         self._state.mark_ready()
         return True
+
+    async def bootstrap(
+        self,
+        config: DeploymentBootstrapConfig,
+        finisher: Callable[[bool, Optional[str]], Awaitable[None]]
+    ) -> None:
+
+        assert config.hostname
+        assert config.address
+        assert config.token
+
+        hostname = config.hostname
+        address = config.address
+        token = config.token
+
+        if self._state.error:
+            raise NodeCantBootstrapError("node is in error state")
+
+        try:
+            await self._prepare_bootstrap(hostname, address, token)
+        except NodeError as e:
+            logger.error(f"bootstrap prepare error: {e.message}")
+            raise e
+
+        assert not self._bootstrapper
+        await self._start_bootstrap()
+        self._bootstrapper = Bootstrap()
+
+        async def finish_bootstrap_cb(
+            success: bool,
+            error: Optional[str]
+        ) -> None:
+            if not success:
+                logger.error(f"bootstrap finish error: {error}")
+                assert self._state.bootstrapping
+                self._state.mark_error()
+
+            await finisher(success, error)
+
+        try:
+            await self._bootstrapper.bootstrap(
+                address, finish_bootstrap_cb
+            )
+        except BootstrapError as e:
+            logger.error(f"bootstrap error: {e.message}")
+            raise NodeCantBootstrapError(e.message)
+
+    async def _prepare_bootstrap(
```

Comment:
(nit since this is carried over) I'd really like to get rid of "prepare bootstrap". Bootstrapping is preparing, so this is like "bootstrap bootstrap" and so it bothers me (totally a nit). It could just as well be "prepare_etcd"?

Reply:
I get that, but _bootstrap_ is also an action. You can prepare to start said action, start said action, and then finish said action. Aside from putting all the functions in the bootstrap function's context, I don't know a decent way out of this.

Reply:
why not just call it prepare_etcd?

(again, this is a nit, so I'm not really fussed)

Reply:
works for this function, sure; but I was addressing the other comments as well ;)

Reply:
Ah yes, in those cases I had just suggested rolling them into the one bootstrap.

Reply:
I guess they can be moved under `bootstrap()`'s context :shrug: